File: combinatorics/resource/ST/project_5/bai_toan_4/mutil_graph/am_to_eal.cpp
#include <iostream>
#include <vector>

using namespace std;

// Ham chuyen AM sang EAL cho da do thi
vector<vector<pair<int, int>>> amToEAL(const vector<vector<int>>& matrix) {
    int n = matrix.size();
    vector<vector<pair<int, int>>> eal(n); // Tao danh sach rong

    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < n; ++j) {
            int weight = matrix[i][j]; // So canh tu i den j
            if (weight > 0) {
                eal[i].emplace_back(j, weight); // Them (j, weight)
            }
        }
    }

    return eal;
}

int main() {
    int n;
    cout << "Nhap so dinh: ";
    cin >> n;

    vector<vector<int>> matrix(n, vector<int>(n));
    cout << "Nhap ma tran ke:\n";
    for (int i = 0; i < n; ++i)
        for (int j = 0; j < n; ++j)
            cin >> matrix[i][j];

    // Thuc hien chuyen doi
    auto eal = amToEAL(matrix);

    cout << "Danh sach ke mo rong (EAL):\n";
    for (int i = 0; i < n; ++i) {
        cout << i << ": ";
        for (auto [v, w] : eal[i]) {
            cout << "(" << v << "," << w << ") ";
        }
        cout << endl;
    }

    return 0;
}

File: combinatorics/resource/ST/project_5/bai_toan_4/mutil_graph/am_to_eal.py
# Ham chuyen AM sang EAL cho da do thi
def am_to_eal(matrix):
    n = len(matrix)
    eal = [[] for _ in range(n)]  # Tao danh sach ke rong cho n dinh

    for i in range(n):
        for j in range(n):
            weight = matrix[i][j]  # So canh tu i den j
            if weight > 0:
                eal[i].append((j, weight))  # Them (j, weight) vao danh sach

    return eal

# Nhap du lieu tu ban phim
n = int(input("Nhap so dinh: "))
print("Nhap ma tran ke:")
matrix = []
for _ in range(n):
    row = list(map(int, input().split()))
    matrix.append(row)

# Chuyen doi va in ket qua
eal = am_to_eal(matrix)
print("Danh sach ke mo rong (EAL):")
for i, neighbors in enumerate(eal):
    print(f"{i}: {neighbors}")